src/dbnode/encoding/m3tsz/roundtrip_test.go

```diff
@@ -112,77 +106,100 @@ func validateRoundTrip(t *testing.T, input []ts.Datapoint, intOpt bool) {
 	defer ctx.Close()
 
 	encoder := NewEncoder(testStartTime, nil, intOpt, nil)
-	for j, v := range input {
-		if j == 0 {
-			encoder.Encode(v, xtime.Millisecond, proto.EncodeVarint(10))
-		} else if j == 10 {
-			encoder.Encode(v, xtime.Microsecond, proto.EncodeVarint(60))
-		} else {
-			encoder.Encode(v, xtime.Second, nil)
+	timeUnits := make([]xtime.Unit, 0, len(input))
+	annotations := make([]ts.Annotation, 0, len(input))
+
+	for i, v := range input {
+		timeUnit := xtime.Second
+		if i == 0 {
+			timeUnit = xtime.Millisecond
+		} else if i == 10 {
+			timeUnit = xtime.Microsecond
+		}
+
+		var annotation, annotationCopy ts.Annotation
+		if i < 5 {
+			annotation = ts.Annotation("foo")
+		} else if i == 10 {
+			annotation = ts.Annotation("bar")
+		}
+
+		if annotation != nil {
+			annotationCopy = append(annotationCopy, annotation...)
+		}
+
+		annotations = append(annotations, annotationCopy)
+		timeUnits = append(timeUnits, timeUnit)
+
+		err := encoder.Encode(v, timeUnit, annotation)
+		require.NoError(t, err)
+
+		for j := range annotation {
+			// Invalidate the original annotation to make sure encoder is not holding a reference to it.
+			annotation[j] = '*'
 		}
 	}
+
 	decoder := NewDecoder(intOpt, nil)
 	stream, ok := encoder.Stream(ctx)
 	require.True(t, ok)
 
 	it := decoder.Decode(stream)
 	require.True(t, ok)
 	defer it.Close()
-	var decompressed []ts.Datapoint
-	j := 0
+
+	i := 0
 	for it.Next() {
-		v, _, a := it.Current()
-
-		if j == 0 {
-			s, _ := proto.DecodeVarint(a)
-			require.Equal(t, uint64(10), s)
-		} else if j == 10 {
-			s, _ := proto.DecodeVarint(a)
-			require.Equal(t, uint64(60), s)
-		} else {
-			require.Nil(t, a)
+		v, u, a := it.Current()
+
+		expectedAnnotation := annotations[i]
```

Comment:
nit: rename `annotations` to `expectedAnnotations`

Reply:
I think I'll leave it as is, as the change would be slightly misleading. `annotations` is what was recorded. There is an additional logics (see the `if` below) to skip repeating values which is applied to produce what is the real `expectedAnnotation`.